sessions/2019/january/Session_22.md

```diff
@@ -0,0 +1,135 @@
+
+#### Fixing Logger implementation
+
+Several fixes were made to the DefaultLogger class:
+
+```java
+@Component
+public class DefaultLogger implements Logger {
+    private final String ip;
+
+    DefaultLogger () {
+        InetAddress inetAddress;
+
+        try {
+            inetAddress = InetAddress.getLocalHost();
+        } catch (UnknownHostException e) {
+            inetAddress = null;
+        }
+
+        if (inetAddress != null){
+            ip = inetAddress.toString();
+        } else {
+            ip = "";
+        }
+    }
+
+    @Override
+    public void log(String message) {
+        Timestamp timestamp = new Timestamp(System.currentTimeMillis());
+        long threadId = Thread.currentThread().getId();
+
+        message = timestamp + "  INFO [" + threadId + "] " + ip + ": " + message;
+        System.out.println(message);
+    }
+}
+```
+
+A very important point here is that the system is asking for the IP address (`InetAddress.getLocalHost()`) in the constructor, so is executed once and the value is cached and used wherever is needed and not consulted each time when the method `log()` is called. Also, `null` inspection was made when the IP address of the system running cannot be loaded for some reason, avoiding to be using methods of `null` objects and receive `Null Pointer` exceptions
+
+By other side, `threads` values cannot be cached and they need to be consulted for each `request` made (each request is a `thread`).
+
+**NOTE:** Java application is a single process that have many threads.
+
+#### Spring workflow in github for estimate effort
+
+In order to estimate effort in a sprint (cicle of 1 week), a number between 1 and 34 must be assigned to each task using fibonacci numbers:
+
+- 1,2,3,5,8,13,21,34
+
+Still the definition of this number is something subjective, eventually becomes constant because is groupal and advance according to the grade of difficulty of the tasks. So the idea of hardness is not the same for a work team of experts than for junior programmers, but anyway the level choosen in the team will be preserved for them and advance through the time.
+
+That said, is important to mention that the most important here is not to measure times, but to be constant, it means that if in the current week I made 13 points the most important is that I preserve my level of work in the following weeks becoming eventually better not worst.
+
+There is some rules:
+
+- Each task have a number of required effort
+- The assigned effort number must be written at the beggining of the task title using square brackets `[]`
+- Not more than 1 ticket in progress column at the time
+- The number of the added effort (summing up all the tasks numbers in the spring) not must be less than the last one, must be more. As we're calibrating we could have been estimated our effort bad and so we reajust but in the long time this becomes some good indicator.
```

Comment:
The total number of points in the current sprint shouldn't be lower than the **completed** effort of last sprint. The reason for this is just so that we don't accidentally underestimate how much work we can do and end up taking longer to complete work that could have been done faster with sufficient quality. Feel free to read up on the [Parkinson's law](https://en.wikipedia.org/wiki/Parkinson%27s_law)